src/lib/operators/aggregate_sort.cpp

```diff
@@ -0,0 +1,585 @@
+   *               or in other words: the last entry of a group + 1 row, similar to vector::end()
+   *
+   *               It is (in _aggregate_values) used as indicator that a group has ended just before this row.
+   *               As a result, the start of the very first group is not represented in aggregate_group_pointers,
+   *               as there was no group before that ended.
+   *               Further, the end of the last group is not represented as well.
+   *               This is because no new group starts after it.
+   *               So in total, aggregate_group_pointers will contain one element less than there are groups.
+   */
+  std::set<RowID> aggregate_group_pointers;
+  auto chunks = sorted_table->chunks();
+  for (const auto& column_id : _groupby_column_ids) {
+    auto data_type = input_table->column_data_type(column_id);
+    resolve_data_type(data_type, [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+      std::optional<ColumnDataType> previous_value;
+      ChunkID chunk_id{0};
+      for (const auto& chunk : chunks) {
+        auto segment = chunk->get_segment(column_id);
+        segment_iterate<ColumnDataType>(*segment, [&](const auto& position) {
+          if (previous_value && position.value() != *previous_value) {
+            aggregate_group_pointers.insert(RowID{chunk_id, position.chunk_offset()});
+          }
+          // TODO(anybody) shouldnt it be sufficient to do this in an else-block?
+          // It should work both ways and as far as I know both ways are used in Hyrise
+          // With one way we can save the emplace and with the other way we save the branching.
+          previous_value.emplace(position.value());
+        });
+        chunk_id++;
+      }
+    });
+  }
+
+  /*
+   * For all group by columns
+   *   For each group by column
+   *     output the column value at the start of the group (it is per definition the same in the whole group)
+   * Write outputted values into the result table
+   */
+  size_t groupby_index = 0;
+  for (const auto& column_id : _groupby_column_ids) {
+    auto aggregate_group_pointer_iter = aggregate_group_pointers.cbegin();
+    auto data_type = input_table->column_data_type(column_id);
+    resolve_data_type(data_type, [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+      std::vector<ColumnDataType> values(aggregate_group_pointers.size() + 1);
+      std::vector<bool> null_values(aggregate_group_pointers.size() + 1);
+      size_t value_index = 0;
+      ChunkID chunk_id{0};
+      for (const auto& chunk : chunks) {
+        auto segment = chunk->get_segment(column_id);
+        segment_iterate<ColumnDataType>(*segment, [&](const auto& position) {
+          // ToDo access the row ids directly without iterating
+          // thoughts: its just a small part, but this should reduce runtime for this loop from O(input) to O(output)
```

Comment:
Are you aware that you can pass a PosList into segment_iterate? If you convert your set to one, this should do the job.

Reply:
@mrks are you referring to `segment_iterate_filtered()` from `segment_iterate.hpp` ? 
We were not aware of that, but I'm not really sure how it works either.
IIRC is a PosList a vector of RowIDs, and RowIDs consist of chunk id and chunk offset. Does `segment_iterate_filtered` simply ignore the chunk id?
The set (`group_boundaries`) contains RowIDs spread over several chunks, so we'll probably have to create a PosList for each chunk/segment, from the corresponding subset of `group_boundaries`.

Reply:
Yes, it ignores the chunk ID. I missed the part of being chunk IDs being from different chunks; you certainly don't want to split this. Maybe add a comment for this?